File: src/系统/03．技能系统/00．技能模板+函数/01．技能函数/06．施法·蓄力·充能/进度条特效.lua
local ____lualib = require("lualib_bundle")
local Map = ____lualib.Map
local __TS__New = ____lualib.__TS__New
local __TS__Iterator = ____lualib.__TS__Iterator
local __TS__ArraySort = ____lualib.__TS__ArraySort
local ____exports = {}
local _____53D6_53E5_67C4ID, _____5C1D_8BD5_505C_6B62_8FDB_5EA6_6761_8DDF_968F_9A71_52A8, _____79FB_9664_8FDB_5EA6_6761_7279_6548, removePeriodicCallback, GetHandleId, DestroyEffect, DzSetEffectScale, _____8FDB_5EA6_6761_6620_5C04, _____5355_4F4D_8FDB_5EA6_6761_6620_5C04, _____8DDF_968F_56DE_8C03ID
function _____53D6_53E5_67C4ID(h)
    if h == nil or h == 0 then
        return 0
    end
    return GetHandleId(h)
end
function _____5C1D_8BD5_505C_6B62_8FDB_5EA6_6761_8DDF_968F_9A71_52A8()
    if _____8FDB_5EA6_6761_6620_5C04.size > 0 or _____8DDF_968F_56DE_8C03ID == 0 then
        return
    end
    removePeriodicCallback(_____8DDF_968F_56DE_8C03ID)
    _____8DDF_968F_56DE_8C03ID = 0
end
function _____79FB_9664_8FDB_5EA6_6761_7279_6548(_____8FDB_5EA6_6761_7279_6548)
    if _____8FDB_5EA6_6761_7279_6548 == nil or _____8FDB_5EA6_6761_7279_6548 == 0 then
        return
    end
    local _____8FDB_5EA6_6761_7279_6548ID = _____53D6_53E5_67C4ID(_____8FDB_5EA6_6761_7279_6548)
    local _____6570_636E = _____8FDB_5EA6_6761_6620_5C04:get(_____8FDB_5EA6_6761_7279_6548ID)
    if _____6570_636E ~= nil then
        _____5355_4F4D_8FDB_5EA6_6761_6620_5C04:delete(_____6570_636E["跟随单位ID"])
    end
    _____8FDB_5EA6_6761_6620_5C04:delete(_____8FDB_5EA6_6761_7279_6548ID)
    DzSetEffectScale(_____8FDB_5EA6_6761_7279_6548, 0)
    DestroyEffect(_____8FDB_5EA6_6761_7279_6548)
    _____5C1D_8BD5_505C_6B62_8FDB_5EA6_6761_8DDF_968F_9A71_52A8()
end
--- 进度条特效模块（施法进度条）
-- 
-- 说明：
-- 1. 直接创建独立点特效，不绑定单位或附着点
-- 2. 进度条颜色、动画速度、动画序号都通过特效接口控制
-- 3. 由中心计时器每 0.03 秒刷新到单位坐标与飞行高度之上
local jass = require("jass.common")
local japi = require("jass.japi")
local ____require_result_0 = require("lib.扩展函数.Star扩展函数.04．EC扩展库")
local EC_CreateEffect = ____require_result_0.EC_CreateEffect
local ____require_result_1 = require("系统.00．核心系统.05．中心计时器")
local addPeriodicCallback = ____require_result_1.addPeriodicCallback
removePeriodicCallback = ____require_result_1.removePeriodicCallback
local PROGRESSBAR_MODEL = "war3mapImported\\Progressbar.mdx"
____exports["默认进度条高度偏移"] = 233
local FOLLOW_INTERVAL_MS = 30
local DEFAULT_SCALE = 1.5
local DEFAULT_ANIM_INDEX = 0
local DEFAULT_COLOR_RGBA = {r = 255, g = 255, b = 0, a = 255}
local UNIT_ALIVE_LIFE = 0.405
GetHandleId = jass.GetHandleId
local GetUnitX = jass.GetUnitX
local GetUnitY = jass.GetUnitY
local GetUnitFlyHeight = jass.GetUnitFlyHeight
local GetUnitTypeId = jass.GetUnitTypeId
local GetUnitState = jass.GetUnitState
local IsUnitType = jass.IsUnitType
DestroyEffect = jass.DestroyEffect
DzSetEffectScale = japi.DzSetEffectScale
local DzSetEffectAnimation = japi.DzSetEffectAnimation
local EXSetEffectSpeed = japi.EXSetEffectSpeed
local EXSetEffectXY = japi.EXSetEffectXY
local EXSetEffectZ = japi.EXSetEffectZ
local DzSetEffectVertexColor = japi.DzSetEffectVertexColor
local DzGetColor = japi.DzGetColor
_____8FDB_5EA6_6761_6620_5C04 = __TS__New(Map)
_____5355_4F4D_8FDB_5EA6_6761_6620_5C04 = __TS__New(Map)
_____8DDF_968F_56DE_8C03ID = 0
local function _____83B7_53D6_6709_5E8F_8FDB_5EA6_6761_7279_6548ID_5217_8868()
    local ids = {}
    for ____, id in __TS__Iterator(_____8FDB_5EA6_6761_6620_5C04:keys()) do
        ids[#ids + 1] = id
    end
    __TS__ArraySort(
        ids,
        function(____, a, b) return a - b end
    )
    return ids
end
local function _____5355_4F4D_5B58_6D3B(u)
    if u == nil or u == 0 then
        return false
    end
    if GetUnitTypeId(u) == 0 then
        return false
    end
    if IsUnitType(u, jass.UNIT_TYPE_DEAD) then
        return false
    end
    return GetUnitState(u, jass.UNIT_STATE_LIFE) > UNIT_ALIVE_LIFE
end
local function _____5237_65B0_8FDB_5EA6_6761_4F4D_7F6E(_____6570_636E)
    EXSetEffectXY(
        _____6570_636E["进度条特效"],
        GetUnitX(_____6570_636E["跟随单位"]),
        GetUnitY(_____6570_636E["跟随单位"])
    )
    EXSetEffectZ(
        _____6570_636E["进度条特效"],
        GetUnitFlyHeight(_____6570_636E["跟随单位"]) + _____6570_636E["高度偏移"]
    )
end
local function _____5237_65B0_6240_6709_8FDB_5EA6_6761_4F4D_7F6E()
    local ids = _____83B7_53D6_6709_5E8F_8FDB_5EA6_6761_7279_6548ID_5217_8868()
    do
        local i = 0
        while i < #ids do
            do
                local _____6570_636E = _____8FDB_5EA6_6761_6620_5C04:get(ids[i + 1])
                if _____6570_636E == nil then
                    goto __continue15
                end
                if not _____5355_4F4D_5B58_6D3B(_____6570_636E["跟随单位"]) then
                    _____79FB_9664_8FDB_5EA6_6761_7279_6548(_____6570_636E["进度条特效"])
                    goto __continue15
                end
                _____5237_65B0_8FDB_5EA6_6761_4F4D_7F6E(_____6570_636E)
            end
            ::__continue15::
            i = i + 1
        end
    end
end
local function _____786E_4FDD_8FDB_5EA6_6761_8DDF_968F_9A71_52A8()
    if _____8DDF_968F_56DE_8C03ID ~= 0 then
        return
    end
    _____8DDF_968F_56DE_8C03ID = addPeriodicCallback(FOLLOW_INTERVAL_MS, _____5237_65B0_6240_6709_8FDB_5EA6_6761_4F4D_7F6E)
end
local function _____88C1_526A_5230_5B57_8282(value)
    if value <= 0 then
        return 0
    end
    if value >= 255 then
        return 255
    end
    return jass.R2I(value)
end
____exports["创建进度条特效"] = function(_____5355_4F4D, _____9009_9879)
    if not _____5355_4F4D_5B58_6D3B(_____5355_4F4D) then
        return nil
    end
    local _____5355_4F4DID = _____53D6_53E5_67C4ID(_____5355_4F4D)
    if _____5355_4F4DID == 0 then
        return nil
    end
    local _____5DF2_6709_8FDB_5EA6_6761 = _____5355_4F4D_8FDB_5EA6_6761_6620_5C04:get(_____5355_4F4DID)
    if _____5DF2_6709_8FDB_5EA6_6761 ~= nil then
        _____79FB_9664_8FDB_5EA6_6761_7279_6548(_____5DF2_6709_8FDB_5EA6_6761)
    end
    local _____7F29_653E = _____9009_9879 and _____9009_9879["缩放"] or DEFAULT_SCALE
    local _____52A8_753B_5E8F_53F7 = _____9009_9879 and _____9009_9879["动画序号"] or DEFAULT_ANIM_INDEX
    local _____52A8_753B_901F_5EA6 = _____9009_9879 and _____9009_9879["动画速度"]
    local _____989C_8272 = _____9009_9879 and _____9009_9879["颜色"] or DEFAULT_COLOR_RGBA
    local _____9AD8_5EA6_504F_79FB = _____9009_9879 and _____9009_9879["高度偏移"] or ____exports["默认进度条高度偏移"]
    local x = GetUnitX(_____5355_4F4D)
    local y = GetUnitY(_____5355_4F4D)
    local z = GetUnitFlyHeight(_____5355_4F4D) + _____9AD8_5EA6_504F_79FB
    local _____8FDB_5EA6_6761_7279_6548 = EC_CreateEffect(
        PROGRESSBAR_MODEL,
        x,
        y,
        z,
        0,
        _____7F29_653E,
        _____52A8_753B_901F_5EA6 or 1,
        -1
    )
    if _____8FDB_5EA6_6761_7279_6548 == nil or _____8FDB_5EA6_6761_7279_6548 == 0 then
        return nil
    end
    DzSetEffectVertexColor(
        _____8FDB_5EA6_6761_7279_6548,
        DzGetColor(
            _____88C1_526A_5230_5B57_8282(_____989C_8272.a),
            _____88C1_526A_5230_5B57_8282(_____989C_8272.r),
            _____88C1_526A_5230_5B57_8282(_____989C_8272.g),
            _____88C1_526A_5230_5B57_8282(_____989C_8272.b)
        )
    )
    DzSetEffectAnimation(_____8FDB_5EA6_6761_7279_6548, _____52A8_753B_5E8F_53F7, 0)
    EXSetEffectSpeed(_____8FDB_5EA6_6761_7279_6548, _____52A8_753B_901F_5EA6 or 1)
    local _____6570_636E = {["进度条特效"] = _____8FDB_5EA6_6761_7279_6548, ["跟随单位"] = _____5355_4F4D, ["跟随单位ID"] = _____5355_4F4DID, ["高度偏移"] = _____9AD8_5EA6_504F_79FB}
    _____8FDB_5EA6_6761_6620_5C04:set(
        _____53D6_53E5_67C4ID(_____8FDB_5EA6_6761_7279_6548),
        _____6570_636E
    )
    _____5355_4F4D_8FDB_5EA6_6761_6620_5C04:set(_____5355_4F4DID, _____8FDB_5EA6_6761_7279_6548)
    _____5237_65B0_8FDB_5EA6_6761_4F4D_7F6E(_____6570_636E)
    _____786E_4FDD_8FDB_5EA6_6761_8DDF_968F_9A71_52A8()
    return _____8FDB_5EA6_6761_7279_6548
end
____exports["销毁进度条特效"] = function(_____8FDB_5EA6_6761_5355_4F4D)
    _____79FB_9664_8FDB_5EA6_6761_7279_6548(_____8FDB_5EA6_6761_5355_4F4D)
end
____exports["销毁单位进度条特效"] = function(_____5355_4F4D)
    if _____5355_4F4D == nil or _____5355_4F4D == 0 then
        return
    end
    local _____5355_4F4DID = _____53D6_53E5_67C4ID(_____5355_4F4D)
    if _____5355_4F4DID == 0 then
        return
    end
    local _____8FDB_5EA6_6761_5355_4F4D = _____5355_4F4D_8FDB_5EA6_6761_6620_5C04:get(_____5355_4F4DID)
    if _____8FDB_5EA6_6761_5355_4F4D ~= nil then
        _____79FB_9664_8FDB_5EA6_6761_7279_6548(_____8FDB_5EA6_6761_5355_4F4D)
    end
end
____exports["是否存在进度条特效"] = function(_____5355_4F4D)
    if _____5355_4F4D == nil or _____5355_4F4D == 0 then
        return false
    end
    return _____5355_4F4D_8FDB_5EA6_6761_6620_5C04:has(_____53D6_53E5_67C4ID(_____5355_4F4D))
end
____exports["获取单位进度条特效"] = function(_____5355_4F4D)
    if _____5355_4F4D == nil or _____5355_4F4D == 0 then
        return nil
    end
    return _____5355_4F4D_8FDB_5EA6_6761_6620_5C04:get(_____53D6_53E5_67C4ID(_____5355_4F4D))
end
____exports["清除所有进度条特效"] = function()
    local _____8FDB_5EA6_6761_7279_6548ID_5217_8868 = _____83B7_53D6_6709_5E8F_8FDB_5EA6_6761_7279_6548ID_5217_8868()
    do
        local i = 0
        while i < #_____8FDB_5EA6_6761_7279_6548ID_5217_8868 do
            local _____6570_636E = _____8FDB_5EA6_6761_6620_5C04:get(_____8FDB_5EA6_6761_7279_6548ID_5217_8868[i + 1])
            if _____6570_636E ~= nil then
                DzSetEffectScale(_____6570_636E["进度条特效"], 0)
                DestroyEffect(_____6570_636E["进度条特效"])
            end
            i = i + 1
        end
    end
    _____8FDB_5EA6_6761_6620_5C04:clear()
    _____5355_4F4D_8FDB_5EA6_6761_6620_5C04:clear()
    _____5C1D_8BD5_505C_6B62_8FDB_5EA6_6761_8DDF_968F_9A71_52A8()
end
local g = _G
if type(g["创建进度条特效"]) ~= "function" then
    g["创建进度条特效"] = ____exports["创建进度条特效"]
end
if type(g["销毁单位进度条特效"]) ~= "function" then
    g["销毁单位进度条特效"] = ____exports["销毁单位进度条特效"]
end
return ____exports

File: TS/系统/03．技能系统/00．技能模板+函数/01．技能函数/06．施法·蓄力·充能/进度条特效.ts
/** @noSelfInFile */
/**
 * 进度条特效模块（施法进度条）
 *
 * 说明：
 * 1. 直接创建独立点特效，不绑定单位或附着点
 * 2. 进度条颜色、动画速度、动画序号都通过特效接口控制
 * 3. 由中心计时器每 0.03 秒刷新到单位坐标与飞行高度之上
 */

const jass = require("jass.common") as any;
const japi = require("jass.japi") as any;

const { EC_CreateEffect } = require("lib.扩展函数.Star扩展函数.04．EC扩展库") as {
  EC_CreateEffect: (this: void, path: string, x: number, y: number, z: number, fac: number, size: number, speed: number, time: number) => any;
};
const { addPeriodicCallback, removePeriodicCallback } = require("系统.00．核心系统.05．中心计时器") as {
  addPeriodicCallback: (this: void, intervalMs: number, callback: (this: void, variable?: any) => void, variable?: any) => number;
  removePeriodicCallback: (this: void, id: number) => void;
};

const PROGRESSBAR_MODEL = "war3mapImported\\Progressbar.mdx";
export const 默认进度条高度偏移 = 233.0;
const FOLLOW_INTERVAL_MS = 30;
const DEFAULT_SCALE = 1.5;
const DEFAULT_ANIM_INDEX = 0;
const DEFAULT_COLOR_RGBA = { r: 255, g: 255, b: 0, a: 255 };
const UNIT_ALIVE_LIFE = 0.405;

const GetHandleId = jass.GetHandleId as (h: any) => number;
const GetUnitX = jass.GetUnitX as (u: any) => number;
const GetUnitY = jass.GetUnitY as (u: any) => number;
const GetUnitFlyHeight = jass.GetUnitFlyHeight as (u: any) => number;
const GetUnitTypeId = jass.GetUnitTypeId as (u: any) => number;
const GetUnitState = jass.GetUnitState as (u: any, state: any) => number;
const IsUnitType = jass.IsUnitType as (u: any, whichType: any) => boolean;
const DestroyEffect = jass.DestroyEffect as (effect: any) => void;
const DzSetEffectScale = japi.DzSetEffectScale as (effect: any, scale: number) => void;
const DzSetEffectAnimation = japi.DzSetEffectAnimation as (effect: any, animationIndex: number, flag: number) => void;
const EXSetEffectSpeed = japi.EXSetEffectSpeed as (effect: any, speed: number) => void;
const EXSetEffectXY = japi.EXSetEffectXY as (effect: any, x: number, y: number) => void;
const EXSetEffectZ = japi.EXSetEffectZ as (effect: any, z: number) => void;
const DzSetEffectVertexColor = japi.DzSetEffectVertexColor as (effect: any, color: number) => void;
const DzGetColor = japi.DzGetColor as (alpha: number, red: number, green: number, blue: number) => number;

export interface 进度条特效选项 {
  高度偏移?: number;
  缩放?: number;
  动画序号?: number;
  动画速度?: number;
  颜色?: { r: number; g: number; b: number; a: number };
}

interface 进度条特效数据 {
  进度条特效: any;
  跟随单位: any;
  跟随单位ID: number;
  高度偏移: number;
}

const 进度条映射 = new Map<number, 进度条特效数据>();
const 单位进度条映射 = new Map<number, any>();
let 跟随回调ID = 0;

function 取句柄ID(h: any): number {
  if (h == null || h === 0) return 0;
  return GetHandleId(h);
}

function 获取有序进度条特效ID列表(): number[] {
  const ids: number[] = [];
  for (const id of 进度条映射.keys()) {
    ids.push(id);
  }
  ids.sort((a, b) => a - b);
  return ids;
}

function 单位存活(u: any): boolean {
  if (u == null || u === 0) return false;
  if (GetUnitTypeId(u) === 0) return false;
  if (IsUnitType(u, jass.UNIT_TYPE_DEAD)) return false;
  return GetUnitState(u, jass.UNIT_STATE_LIFE) > UNIT_ALIVE_LIFE;
}

function 刷新进度条位置(this: void, 数据: 进度条特效数据): void {
  EXSetEffectXY(数据.进度条特效, GetUnitX(数据.跟随单位), GetUnitY(数据.跟随单位));
  EXSetEffectZ(数据.进度条特效, GetUnitFlyHeight(数据.跟随单位) + 数据.高度偏移);
}

function 刷新所有进度条位置(this: void): void {
  const ids = 获取有序进度条特效ID列表();
  for (let i = 0; i < ids.length; i++) {
    const 数据 = 进度条映射.get(ids[i]);
    if (数据 == null) continue;
    if (!单位存活(数据.跟随单位)) {
      移除进度条特效(数据.进度条特效);
      continue;
    }
    刷新进度条位置(数据);
  }
}

function 确保进度条跟随驱动(this: void): void {
  if (跟随回调ID !== 0) return;
  跟随回调ID = addPeriodicCallback(FOLLOW_INTERVAL_MS, 刷新所有进度条位置);
}

function 尝试停止进度条跟随驱动(this: void): void {
  if (进度条映射.size > 0 || 跟随回调ID === 0) return;
  removePeriodicCallback(跟随回调ID);
  跟随回调ID = 0;
}

function 裁剪到字节(value: number): number {
  if (value <= 0) return 0;
  if (value >= 255) return 255;
  return jass.R2I(value);
}

function 移除进度条特效(进度条特效: any): void {
  if (进度条特效 == null || 进度条特效 === 0) return;

  const 进度条特效ID = 取句柄ID(进度条特效);
  const 数据 = 进度条映射.get(进度条特效ID);
  if (数据 != null) {
    单位进度条映射.delete(数据.跟随单位ID);
  }

  进度条映射.delete(进度条特效ID);
  DzSetEffectScale(进度条特效, 0);
  DestroyEffect(进度条特效);
  尝试停止进度条跟随驱动();
}

export function 创建进度条特效(单位: any, 选项?: 进度条特效选项): any {
  if (!单位存活(单位)) return null;

  const 单位ID = 取句柄ID(单位);
  if (单位ID === 0) return null;

  const 已有进度条 = 单位进度条映射.get(单位ID);
  if (已有进度条 != null) {
    移除进度条特效(已有进度条);
  }

  const 缩放 = 选项?.缩放 ?? DEFAULT_SCALE;
  const 动画序号 = 选项?.动画序号 ?? DEFAULT_ANIM_INDEX;
  const 动画速度 = 选项?.动画速度;
  const 颜色 = 选项?.颜色 ?? DEFAULT_COLOR_RGBA;
  const 高度偏移 = 选项?.高度偏移 ?? 默认进度条高度偏移;
  const x = GetUnitX(单位);
  const y = GetUnitY(单位);
  const z = GetUnitFlyHeight(单位) + 高度偏移;
  const 进度条特效 = EC_CreateEffect(PROGRESSBAR_MODEL, x, y, z, 0, 缩放, 动画速度 ?? 1, -1);
  if (进度条特效 == null || 进度条特效 === 0) return null;
  DzSetEffectVertexColor(进度条特效, DzGetColor(
    裁剪到字节(颜色.a),
    裁剪到字节(颜色.r),
    裁剪到字节(颜色.g),
    裁剪到字节(颜色.b),
  ));
  // 进度条动画的原始周期为1秒：持续5秒使用0.2，持续3秒使用1/3。
  DzSetEffectAnimation(进度条特效, 动画序号, 0);
  EXSetEffectSpeed(进度条特效, 动画速度 ?? 1);

  const 数据: 进度条特效数据 = {
    进度条特效,
    跟随单位: 单位,
    跟随单位ID: 单位ID,
    高度偏移,
  };

  进度条映射.set(取句柄ID(进度条特效), 数据);
  单位进度条映射.set(单位ID, 进度条特效);
  刷新进度条位置(数据);
  确保进度条跟随驱动();

  return 进度条特效;
}

export function 销毁进度条特效(进度条单位: any): void {
  移除进度条特效(进度条单位);
}

export function 销毁单位进度条特效(单位: any): void {
  if (单位 == null || 单位 === 0) return;
  const 单位ID = 取句柄ID(单位);
  if (单位ID === 0) return;

  const 进度条单位 = 单位进度条映射.get(单位ID);
  if (进度条单位 != null) {
    移除进度条特效(进度条单位);
  }

}

export function 是否存在进度条特效(单位: any): boolean {
  if (单位 == null || 单位 === 0) return false;
  return 单位进度条映射.has(取句柄ID(单位));
}

export function 获取单位进度条特效(单位: any): any {
  if (单位 == null || 单位 === 0) return undefined;
  return 单位进度条映射.get(取句柄ID(单位));
}

export function 清除所有进度条特效(): void {
  const 进度条特效ID列表 = 获取有序进度条特效ID列表();
  for (let i = 0; i < 进度条特效ID列表.length; i++) {
    const 数据 = 进度条映射.get(进度条特效ID列表[i]);
    if (数据 != null) {
      DzSetEffectScale(数据.进度条特效, 0);
      DestroyEffect(数据.进度条特效);
    }
  }
  进度条映射.clear();
  单位进度条映射.clear();
  尝试停止进度条跟随驱动();
}

const g = globalThis as any;
if (typeof g.创建进度条特效 !== "function") {
  g.创建进度条特效 = 创建进度条特效;
}
if (typeof g.销毁单位进度条特效 !== "function") {
  g.销毁单位进度条特效 = 销毁单位进度条特效;
}

export {};
